### PR TITLE
Chat approval

### LINKED ISSUE #8063: Optional moderator approval for chat posts from all new users
We want to have a setting to temporarily turn off the ability of new users to post before allowed to do so by a moderator. This would be very helpful in situations where a single user (or several trolls) tries to make many accounts in a row to dodge a ban and spam chat.

Details:
1. All existing users are always allowed to post.
2. Under normal conditions, all new users are allowed as soon as they've created an account (just as they can now) BUT:
3. There should be a flag (an environment variable?) that mods can use to cause every new user account to be unable to post until a mod enables posting for that user. Mods could turn on that flag when a troll keeps creating new accounts. Mods would turn the flag off later when they believed that the troll might have moved on.
4. When a user who is not yet allowed to post tries to post, their message appears in the chat area but visible only to mods (and it's sent to mods in an email). There'd be an icon on the message that mods could use to make the message visible to everyone and to enable normal posting for that user. This behaviour would occur only while the flag was turned on.

EDIT 2021-11-10: One possible solution is in https://github.com/HabitRPG/habitica/pull/12499 but note the comments there from paglias!

### PR DESCRIPTION
**Pull request by TheHollidayInn**
Fixes #8063

### CODE REVIEW

File: website/common/locales/en/groups.json
Comment (Alys) [inline website/common/locales/en/groups.json]: These three new strings could go into website/server/libs/apiMessages.js since they don't need to be translated.

File: website/server/models/group.js
Comment (Alys) [inline website/server/models/group.js]: Should oneDayAgo be given a more generic name in case we change the number of days later? Perhaps something like cutOffDate?

Also should 1 be a constant? That would allow us to reference the constant in the tests - for example, for this line, instead of `2` we could use the constant + 1:
` +    const twoDaysAgo = moment().startOf('day').subtract(2, 'days').toDate();`

It might even be good to use an ENV var instead of a constant so that we can change it easily to cover the case in the second paragraph at https://github.com/HabitRPG/habitica/issues/8063#issuecomment-353176169 (a group of trolls joins a few days in the past then starts spamming).

Although this is making the PR more complex so if you want to just leave it as it is we can add customisation later. :)

File: website/client/components/chat/chatCard.vue
Comment (Alys) [inline website/client/components/chat/chatCard.vue]: I'm not sure we should reuse the `approvalRequired` string. It's used for task approvals in group plans and if requirements ever change for that, we'd want to be able to change the text without affecting chat message approvals. Maybe the string could be `approvalIsRequired` for consistency with the `approvalNotRequired` string?

If that is changed, it might also be helpful for future coding if the functions and variables called `approvalRequired` are changed to be the same as the string, so that we can git grep for that one phrase and find all the relevant code, without the group task approval code appearing too. But that's a pretty minor point so ignore it if you wish. :)

File: website/server/models/group.js
Comment (Alys) [inline website/server/models/group.js]: This should be `newChatMesage` not `newMessage`. With `newMessage` the `true` value is never assigned to the `approvalRequired` field in the message. 

Also that field needs to be added to website/server/models/chat.js

File: website/server/controllers/api-v3/chat.js
Comment (Alys) [inline website/server/controllers/api-v3/chat.js]: This is throwing `ReferenceError: find is not defined` when a mod approves a message.

File: website/client/components/chat/chatMessages.vue
Comment (Alys) [inline website/client/components/chat/chatMessages.vue:142]: I'm checking with the other mods to see if we also want to let the user see their own message. That wasn't specified in the issue but I've realised from my testing that it's confusing when you make a post and don't see it appear in the chat.

File: website/server/controllers/api-v3/chat.js
Comment (Alys) [inline website/server/controllers/api-v3/chat.js]: Should that URL be replaced with https://github.com/HabitRPG/habitica/blob/develop/website/server/models/chat.js now?

File: config.json.example
Comment (Alys) [inline config.json.example]: Is there any way we could turn this feature on and off using a setting in the Hall? That would allow moderators to manage it fully. If that's possible but difficult, it wouldn't need to be done in this PR.

I guess it would pretty much need to be a setting in the database, which would complicate things. A new collection? A value in the Tavern's document? I suppose we don't really have support for this sort of thing at the moment.

File: website/common/locales/en/groups.json
Comment (TheHollidayInn) [reply website/common/locales/en/groups.json]: I can move the server side one. The other two are on the client so we don't have access to that lib there.

File: website/client/components/chat/chatCard.vue
Comment (TheHollidayInn) [reply website/client/components/chat/chatCard.vue]: I can change it to approvalIsRequired. Not really sure about the functions. Seems like organizing the code into the chat libs would be better, and I'll probably skip that to not create more work.

File: config.json.example
Comment (TheHollidayInn) [reply config.json.example]: Yea, we don't have any good support for this.

File: website/common/locales/en/groups.json
Comment (Alys) [reply website/common/locales/en/groups.json]: For those two client-side strings, it might be best to have them hard-coded into the vue files. That would save the translators translating them unnecessarily and would slightly reduce the volume in Transifex. Alternatively, perhaps we could have a `doNotTranslate.json` locales file that's never uploaded to Transifex.

File: website/client/components/chat/chatMessages.vue
Comment (Alys) [reply website/client/components/chat/chatMessages.vue:142]: We've decided that the user should be able to see their own post before it's approved, and should see a message on it indicating that it's pending approval. That will avoid confusion about why their post doesn't appear / why no one is replying and will hopefully prevent them posting the same message over and over again to get attention. Details are here in the issue: https://github.com/HabitRPG/habitica/issues/8063#issuecomment-385624004

**Review by Alys — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
